soroban-env-host/src/vm.rs

```diff
@@ -148,41 +153,41 @@ impl Vm {
             //    early is preferred from resource usage perspective.
             // 3. If a contract contains a call to an non-existent host
             //    function, the current (correct) behavior is to return
-            //    `Wasmi::LinkerError::MissingDefinition` error (which gets
+            //    `Wasmi::errors::LinkerError::MissingDefinition` error (which gets
             //    converted to a `(WasmVm, InvalidAction)`). If that host
             //    function is defined in a later protocol, and we replay that
-            //    contract (in the earlier protocol where it belongs), not
-            //    linking the function preserves the right behavior and error
-            //    code.
+            //    contract (in the earlier protocol where it belongs), we need
+            //    to return the same error.
             let _span0 = tracy_span!("define host functions");
             let ledger_proto = host.with_ledger_info(|li| Ok(li.protocol_version))?;
-            for hf in HOST_FUNCTIONS {
-                if let Some(min_proto) = hf.min_proto {
-                    if parsed_module.proto_version < min_proto || ledger_proto < min_proto {
-                        // We skip linking this hf instead of returning an error
-                        // because we have to support old contracts during replay.
+            parsed_module.with_import_symbols(|module_symbols| {
+                for hf in HOST_FUNCTIONS {
+                    if !module_symbols.contains(&(hf.mod_str, hf.fn_str)) {
                         continue;
                     }
-                }
-                if let Some(max_proto) = hf.max_proto {
-                    if parsed_module.proto_version > max_proto || ledger_proto > max_proto {
-                        // We skip linking this hf instead of returning an error
-                        // because we have to support old contracts during replay.
-                        continue;
+                    if let Some(min_proto) = hf.min_proto {
```

Comment:
@jayz22 I would appreciate your review of this section. I note you explicitly say _not_ to return an error here in the previous code, but unfortunately we really can't "omit the symbols" in the new scheme, since multiple contracts share a single linker. I have tried to emulate the error code correctly but I don't understand why the comment before suggested that returning an error wouldn't work.

Reply:
IIUC we need to have indistinguishable behavior between host functions that don't exist and host functions that are not supported yet by the protocol. What would happen if I upload a contract that imports a secp host function today? (it's not yet present in the env) If we return an error during VM instantiation, then I guess we're good.

Reply:
I believe you've handled it correctly here and there won't be an observable difference. 

But let me elaborate on my code comment with a slightly contrived example. 
In the protocol today (p20), if a contract contains a call to a non-existent host function, `{"c", "3"}` (i.e. secp256r1_verify, available in p21). What the p20 Vm does is:
1. for every existing host function, define it (`linker.define(func)`). There could be an error arriving from `linker.define` call, such as duplicate definition, missing/invalid type etc. 
2. `linker.instantiate`. Attempting to link `{"c", "3"}` in p20, arrives at a `LinkerError::MissingDefinition` in this step. 

Imagine a function in the future protocol replays this contract. It sees the contract attempting to link a host function non-existent to it, in step 1 above, and throws an error immediately. We could potentially be modifying the contract's expected behavior. What if the contract actually failed due to `linker.define` of an existing function in step 1 (that would be have an order later than `{"c", "3"}`), we would be hijacking that by throwing a different error as soon as we saw `{"c", "3"}`. 

This case cannot happen today for p20, because all linker errors are treated as `(WasmVm, InvalidAction)`. Thus we can't be changing observable result. 

My comment above was trying to emphasize that by skipping linking (`continue` instead of `Err`), it was the most safe/future-proof way to ensure the linker preserves the expected protocol behavior.

Reply:
@dmkozh if you upload a function today that is not supported in v21, you get a linker error from wasmi concerning a missing definition during the trial instantiation, which turns into (WasmVm, InvalidAction) error code.

in this new code, the explicit branch that tests the "calling a function not yet supported" condition on line 169 will fail, and return a (WasmVm, InvalidAction).

There are tests for all these sorts of host-vs-contract-vs-protocol ordering cases in `test_host_protocol_gating_for_wasm` that @jayz22 wrote before. This new code passes them all (produces the same results).

The only difference -- as far as I intended and can tell, please point to a way this might be wrong! -- is that the possibility of _registering host functions with the linker_ used to occur inside this loop, whereas now it occurs before the loop altogether (when constructing the shared linker). But I think this is ok: registering host functions shouldn't depend in any way on user input, it's driven by a static list of host functions compiled-in to the env and it should always succeed (as it does in all the tests we run here). If it fails it's an internal error, a bug in the env or something, I don't think we need to consider "the possibility of a change in the timing of an internal error that, if it ever happened, would be based on solely static data" to be an "observable change".

(And if this error _did_ occur in p20, and someone _did_ somehow figure out a way to observe it, I think we'd discover the divergence during the sealed-replay phase of the planned upgrade and we could just mark it as a special case. But since it doesn't fail in practice, as far as we know, even this shouldn't be possible).

Reply:
Agree that no observable difference should be able to occur due to rearranging the order where host functions are registered. Host function registration (`linker.define` in the old code) is just inserting a static list of host functions and checking duplication. There is no chance for it to fail unless there is an internal bug. And we have taken measures to ensure (statically) that host function definitions can not duplicate.